### DISCUSSION TITLE
Insert CVE details into a notes page

### DISCUSSION THREAD

**Discussion by andytinkham** [SilentVoid13/Templater#1031] [Templates Showcase]
I work in software application security, and a lot of my work involves taking known vulnerabilities in the libraries we use and investigating whether the code my team writes is vulnerable to those issues. There is a central database (the NVD) that stores all of these that the National Institute of Standards and Technology stores that has details of all of these issues, along with links to the official announcements from the maintainers, what versions of what software are affected, and things like that.

I want to capture notes to avoid having to re-evaluate whether we're vulnerable every time a particular issue comes up, so I made this set of templates & script: https://gist.github.com/andytinkham/7a98cdca9e34beab75b8d4cb7ea459c6

In the gist are 4 files. 

The first is CVE Details.md. This is just a single-line template file. Sometimes our scanners detect issues before the NVD is updated to have the record. If I come back to an issue's notes page later, I can choose "Templater: Open Insert Template Modal" from the Command Palette and use this to add details later.

The second, CVE-2021-27101.md is an example of a page generated from the template. You can see the original NVD page of data here: https://nvd.nist.gov/vuln/detail/CVE-2021-27101.

The third is my main template (minus a couple of work-specific sections that are just standard markdown).  I have Quick Add set up so I can just open the Command Palette, select "Quick Add: New CVE" and have it make the new note from the template. It prompts for the CVE ID (something like CVE-2021-3711), and then generates the call. This includes going out to NIST's API, retrieving the JSON data structure of CVE details and parsing it all into markdown.  The template also generates links to the NVD and Mitre's CVE database and then includes pre-defined sections to track what I do to investigate whether our different software pieces are vulnerable and to track which repositories or containers have the issue show up. I put some placeholder tags in the metadata, but haven't done much with these yet. 

Both the previous files go in my standard templates directory. (In my case, /Templates).

The final file, cve_details.js, lives in the script directory (/_scripts, for me). This function queries the NIST web service for the CVE details and parses the JSON, including iterating over the variable number of reference links, affected configurations, and common weaknesses (different types of bugs that lead to the security vulnerabilities). The weaknesses are each linked to their own pages so that eventually I can see similar issues. 

I know this is a pretty niche-y script, but hopefully it's helpful to some (even if only as an example). 

Customizations:
- In the script, at the top, there is a constant called language.  NIST's database does occasionally include descriptions and other strings for an issue in other languages, but that's rare. You could change this constant if you wanted to try to get another language but you'll likely get a lot of "undefined" values, unfortunately, because the data just isn't present. I haven't tried to hide these undefined when the data isn't present in other languages.

This is pretty much set now, unless a bug crops up or NIST changes their web service or the JSON schema.

**Comment by andytinkham** [comment]
I just updated the gist with a few more things. If multiple scores are present in the record, they're all listed in the output (which is useful when there's contention on how severe an issue is). I also added some error handling for things I ran into as I inserted the output of this into all my existing CVE notes and made it so even if it fails, it shouldn't leave the obsidian cursor in a spinning state. 

I know of 2 things to do still, but they'll have to wait until tomorrow:
- For issues that aren't on CISA's Known Exploited List, I put a date stamp in the "not currently on the list" message. This date stamp currently uses GMT, so could be a day ahead of the actual date run.
- I found some CVEs that have multi-node configurations but haven't built in anything to handle those yet. Currently, only the first node is included in the output. 

I discovered a few other things in the API that I can add too (like whether there's a US-CERT alert, notes from other sources, and additional data like that, as well as the change history of the CVE). I don't know yet if this will be useful data, but that's on my radar now to investigate further as well.

**Comment by welpdx** [comment]
Wow! As someone getting into Cyber Security, this is awesome!

**Comment by andytinkham** [comment]
Made the last planned set of changes. Everything in the NIST published schema should be parsed and output now, other than the text descriptions of each element in the score vectors (which are redundant with the vector string and available by clicking through to the calculator anyhow). I fixed the multi-node stuff and the time stamp.
